src/CommonProviderImplementation/Helpers.fs

```diff
@@ -302,10 +304,10 @@ module internal ProviderHelpers =
                         providedTypesCache.TryRemove key |> ignore
                     else
                         log (sprintf "Saving generation of type %s for 10 seconds awaiting incremental recreation [%d]" fullTypeName tp.Id)
-                        providedTypesCache.[key] <- CacheValue (providedType, fullKey)
+                        providedTypesCache.AddOrUpdate(key, CacheValue (providedType, fullKey), fun _ _ -> CacheValue (providedType, fullKey)) |> ignore
                         // Remove the cache entry in 10 seconds
                         async { do! Async.Sleep (10000)
-                                providedTypesCache.TryRemove(key) |> ignore } |> Async.StartImmediate
+                                providedTypesCache.TryRemove(key) |> ignore } |> Async.Start
```

Comment:
Because we are using non-UI action we don't have to work in UI thread. We can use TPL.